File: 0216-combination-sum-iii/0216-combination-sum-iii.py
class Solution(object):
    def __init__(self):
        self.result = []
    def backtrack(self, k, n, sum, idx, combo):
        if sum > n:
            return
        if len(combo) == k:
            if sum == n:
                self.result.append(combo[:])
                return
        for i in range(idx, 9 - (k - len(combo)) + 2):
            combo.append(i)
            sum += i
            self.backtrack(k, n, sum, i + 1, combo)
            sum -= i
            combo.pop()
        
    def combinationSum3(self, k, n):
        """
        :type k: int
        :type n: int
        :rtype: List[List[int]]
        """
        self.backtrack(k, n, 0, 1, [])
        return self.result

File: 0216-combination-sum-iii/0216-combination-sum-iii.cpp
class Solution {
private:
    vector<vector<int>> result;
    vector<int> combo;
    void backtrack(int k, int n, int sum, int startIdx){
        if (sum > n) return;
        if (combo.size() == k){
            if (sum == n){
                result.push_back(combo);
                return;
            }
        }
        for (int i = startIdx; i <= 9 - (k - combo.size()) + 1; i++){
            combo.push_back(i);
            sum += i;
            backtrack(k, n, sum, i + 1);
            sum -= i;
            combo.pop_back();
        }
    }
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        backtrack(k, n, 0, 1);
        return result;
    }
};